[pytree][BE] update treespec `is_leaf()` access

XuehaiPan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114947
* #116375
* #116372
* #116374
* __->__ #116371

Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.

XuehaiPan committed:
[pytree] update treespec `is_leaf()` access

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

Skylion007 commented on torch/_functorch/_aot_autograd/utils.py:
```diff
@@ -128,10 +128,10 @@ def set(self, spec):
         ):
             self.is_simple = True
-        if isinstance(self.spec, pytree.LeafSpec):
+        if self.spec.is_leaf():  # type: ignore[union-attr]
```
is_leaf() probably needs to have the return type changed to `typing_extensions.TypeGuard[pytree.LeafSpec]` from bool for there not be an effect on the mypy types. This will change cause `is_leaf()` returning True to act as a type narrowing on the spec assuming that is what you want.

XuehaiPan replied:
`is_leaf` is a method without any positional argument (excluding `self`).

```python
class TreeSpec:
    def is_leaf(self) -> TypeGuard["LeafSpec"]:
        return self.num_nodes == 1 and self.num_leaves == 1
```

`mypy` complains about this:

```text
  Error (MYPY) [valid-type]
    TypeGuard functions must have a positional argument

         566  |        )
         567  |        return self._children_specs
         568  |
    >>>  569  |    def is_leaf(self) -> TypeGuard["LeafSpec"]:
         570  |        return self.num_nodes == 1 and self.num_leaves == 1
```

Although we can have a static function like this:

```python
def is_leaf(treespec: TreeSpec) -> TypeGuard[LeafSpec]:
    return treespec.num_nodes == 1 and treespec.num_leaves == 1
```

Skylion007 replied:
Regardless, this seems like a valid type comment complaint here? 
```suggestion
        if self.spec is not None and self.spec.is_leaf():
```
is what is actually needed here , right?

XuehaiPan replied:
The input argument `spec` should not be `None`. We have `type(self.spec)` above. It will raise `AttributeError` if it is `None`.

XuehaiPan replied:
I updated the method type annotation and added an assert statement above.

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

Skylion007 commented on torch/utils/_pytree.py:
```diff
@@ -510,7 +511,7 @@ def __repr__(self, indent: int = 0) -> str:
         return repr_prefix + repr_suffix
 
-    def is_leaf(self) -> bool:
+    def is_leaf(self) -> TypeGuard["LeafSpec"]:
```
Yeah, so narrow self is explicitly not supported in the TypeGuard PEP. How annoying... Feel free to revert this change then.

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

Skylion007 commented on torch/_functorch/_aot_autograd/utils.py:
```diff
@@ -124,17 +124,18 @@ class PytreeThunk:
             self.is_really_simple = True
 
-    def unflatten(self, x):
+    def unflatten(self, x: List[Any]) -> Any:
```
You could improve with a TypeVar, no?
```suggestion
    def unflatten(self, x: List[_T]) -> _T:
```

Skylion007 replied:
Actually, wow all the flatten / unflatten functions in pytree are really poorly typed. They should try to retain the typing in the iterable if possible like itertools does.

XuehaiPan replied:
The input type is `List[T]`, while the output type can be `T | List[T] | PyTree[T]`. Currently the annotation for `PyTree` is `Any`.

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

zou3519 reviewed (approved):
LGTM, but likely will need a rebase.

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree][BE] update treespec `is_leaf()` access"


Change `isinstance(treespec, LeafSpec) -> treespec.is_leaf()`.



[ghstack-poisoned]

XuehaiPan commented:
@pytorchbot merge
